server/server.go

```diff
@@ -0,0 +1,202 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package xserver
+
+import (
+	"net"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/m3db/m3x/log"
+	"github.com/m3db/m3x/net"
+
+	"github.com/uber-go/tally"
+)
+
+// Server is a server capable of listening to incoming traffic and closing itself
+// when it's shut down.
+type Server interface {
+	// ListenAndServe starts listening to new incoming connections and
+	// handles data from those connections.
+	ListenAndServe() error
+
+	// Close closes the server.
+	Close()
+}
+
+// Handler can handle the data received on connection.
+// It's used in Server once a connection was established.
+type Handler interface {
+	// Handle handles the data received on the connection, this function
+	// should be blocking until the connection is closed or received error.
+	Handle(conn net.Conn)
+}
+
+type serverMetrics struct {
+	openConnections tally.Gauge
+}
+
+func newServerMetrics(scope tally.Scope) serverMetrics {
+	return serverMetrics{
+		openConnections: scope.Gauge("open-connections"),
+	}
+}
+
+type addConnectionFn func(conn net.Conn) bool
+type removeConnectionFn func(conn net.Conn)
+
+type server struct {
+	sync.Mutex
+
+	address  string
+	listener net.Listener
+	opts     Options
+	log      xlog.Logger
+
+	closed   int32
+	numConns int32
+	conns    []net.Conn
+	wgConns  sync.WaitGroup
+	metrics  serverMetrics
+	handler  Handler
+
+	addConnectionFn    addConnectionFn
+	removeConnectionFn removeConnectionFn
+}
+
+// NewServer creates a new server.
+func NewServer(address string, handler Handler, opts Options) Server {
+	instrumentOpts := opts.InstrumentOptions()
+	scope := instrumentOpts.MetricsScope().SubScope("server")
```

Comment:
The convention is usually to assume the scope passed in is the correct scope for the current object.